### PR TITLE
Sketcher: Add concentric capabilities to coincidence constraint.

### PR DESCRIPTION
**Pull request by PaddleStroke**
Enable the coincidence tool to select edges. If those edges are circle, ellipse, arc or arc of ellipse, then it create concentric constraint.

https://github.com/FreeCAD/FreeCAD/issues/7546#issuecomment-1302997618

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [ ]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [ ]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [ ]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7703">pipeline status</a> for feature branch [PR_7703](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7703). Pipeline [685710757](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/685710757) was triggered at [66ca9e6](https://github.com/FreeCAD/FreeCAD/pull/7703/commits/66ca9e65d6558f0b92c925de2f302a0ece3dde1b). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7703">pipeline status</a> for feature branch [PR_7703](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7703). Pipeline [685863484](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/685863484) was triggered at [37564f6](https://github.com/FreeCAD/FreeCAD/pull/7703/commits/37564f6dfe8e66803540c73bdc79f00c51ba1bde). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/Sketcher/Gui/CommandConstraints.cpp
Comment (chennes) [inline src/Mod/Sketcher/Gui/CommandConstraints.cpp:2049]: Just a stylistic comment, you don't have to change it here, but in future code: when using `static_cast`, `dynamic_cast`, etc. where we already explicitly state the type, that's a good place to use `auto` for the variable definition. Sort of a micro-example of the DRY principle.

Comment (PaddleStroke) [reply src/Mod/Sketcher/Gui/CommandConstraints.cpp:2049]: This code has just been moved.
But I take note of your message. I will try to use auto in those situation in the future. 
If I come accross such occurence while making a PR, should I always fix such things along?

Comment (chennes) [reply src/Mod/Sketcher/Gui/CommandConstraints.cpp:2049]: It's hard to make a general statement about it: knowing now that this code was simply relocated, then no, I'd say leave it alone. Switching to `auto` is a minuscule change that makes almost no practical difference, so the very small increase in readability is not worth complicating the review. On the other hand, if you ran across an old-style for loop, and you were already touching that line of code for some other reason, then I'd say yes, let's go ahead and switch to a ranged-for. Brand new code is different, but working with old code IMO there's more wiggle room. There isn't a "policy" either way, we just discuss it in the PR review process and come to a consensus. You are always welcome to push back and tell me why you don't think a particular suggestions makes sense in this particular context: I won't take it personally, we're both trying to make incremental improvements to the code.

File: src/Mod/Sketcher/Gui/CommandConstraints.cpp
Comment (chennes) [reply src/Mod/Sketcher/Gui/CommandConstraints.cpp:2070]: Really, this whole code block has a lot of nearly identical lines. Maybe create a function that takes a `Geometry *` and returns whether it is "concentric-capable"? You could even make it a lambda right inside `applyConstraint`, if it really wasn't needed anywhere else in the code.

File: src/Mod/Sketcher/Gui/CommandConstraints.cpp
Comment (chennes) [inline src/Mod/Sketcher/Gui/CommandConstraints.cpp:1893]: I'd go with something like:
```suggestion
    sToolTipText    = QT_TR_NOOP("Create a coincident constraint between points, or a concentric constraint between circles, arcs, and ellipses");
```

**Review by chennes — CHANGES_REQUESTED** [review]
Overall looks good to me, I like the re-use of the coincident constraint here. A little bit of refactoring to reduce code duplication and this should be good to go.

File: src/Mod/Sketcher/Gui/CommandConstraints.cpp
Comment (chennes) [reply src/Mod/Sketcher/Gui/CommandConstraints.cpp:2049]: It's hard to make a general statement about it: knowing now that this code was simply relocated, then no, I'd say leave it alone. Switching to `auto` is a minuscule change that makes almost no practical difference, so the very small increase in readability is not worth complicating the review. On the other hand, if you ran across an old-style for loop, and you were already touching that line of code for some other reason, then I'd say yes, let's go ahead and switch to a ranged-for. Brand new code is different, but working with old code IMO there's more wiggle room. There isn't a "policy" either way, we just discuss it in the PR review process and come to a consensus. You are always welcome to push back and tell me why you don't think a particular suggestions makes sense in this particular context: I won't take it personally, we're both trying to make incremental improvements to the code.

File: src/Mod/Sketcher/Gui/CommandConstraints.cpp
Comment (PaddleStroke) [reply src/Mod/Sketcher/Gui/CommandConstraints.cpp:2070]: It's also in activated()
I thought about making a function, but wasn't sure where to put the function. 
I thought about lambda in applyConstraint, but then the occurence in activated is left behind.

Comment (PaddleStroke) [reply src/Mod/Sketcher/Gui/CommandConstraints.cpp:2070]: Ok just on top of the file I guess. There are a few other functions like this there.

Comment (chennes) [reply src/Mod/Sketcher/Gui/CommandConstraints.cpp:2070]: Yep - probably in an anonymous namespace, just like the other PR. That will help cut down on what's already a beast of a file!

**Comment by PaddleStroke** [comment]
I just fixed the sentence, and added the function to remove code repeat and forced pushed.

File: src/Mod/Sketcher/Gui/CommandConstraints.cpp
Comment (PaddleStroke) [reply src/Mod/Sketcher/Gui/CommandConstraints.cpp]: Other functions aren't in anonymous namespace. So doing so would break homogeneity.

**Comment by PaddleStroke** [comment]
I forgot your comment for the tooltip. I jsut added it and re-force-pushed.

File: src/Mod/Sketcher/Gui/CommandConstraints.cpp
Comment (PaddleStroke) [reply src/Mod/Sketcher/Gui/CommandConstraints.cpp]: So I'm unsure. What do you think about that?

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7703">pipeline status</a> for feature branch [PR_7703](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7703). Pipeline [685991591](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/685991591) was triggered at [c1c168e](https://github.com/FreeCAD/FreeCAD/pull/7703/commits/c1c168e719db0fd702ac07d50269a9801100b350). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7703">pipeline status</a> for feature branch [PR_7703](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7703). Pipeline [685994285](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/685994285) was triggered at [129011e](https://github.com/FreeCAD/FreeCAD/pull/7703/commits/129011e90eeba1771eca209052d651392d5afe9f). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/Sketcher/Gui/CommandConstraints.cpp
Comment (chennes) [reply src/Mod/Sketcher/Gui/CommandConstraints.cpp]: I'd say don't worry about the anonymous namespace bit for now: it's relatively recent modern C++ coding advice that we should seek to follow when it's reasonable to do so, but don't stress ourselves out over. Someday maybe someone will decide to do a cleanup pass on this file and address some of the linter issues, and they can put all of those functions in an anonymous namespace then. It's not glamorous work, but it's also pretty low-hanging fruit 😄 .

**Review by chennes — APPROVED** [review]
This looks great now, thanks! I'm going to wait for the CI to complete before merging, but it compiles, runs, and behaves correctly on MSVC/Win11.

**Comment by 0penBrain** [comment]
@PaddleStroke This PR introduced severe regression.
To reproduce :
* Open new sketch
* Draw an arc
* Draw a line
* Select both arc (edge) + one of the line endpoints
* Add coincidence constraint --> A malformed constraint is created
This virtually is the case with any conic + any point selected.

**Comment by PaddleStroke** [comment]
Fixed in : https://github.com/FreeCAD/FreeCAD/pull/7794

**Comment by 0penBrain** [comment]
Wiki pages about Coincident constraint and Release notes to be updated

**Comment by PaddleStroke** [comment]
done
https://wiki.freecadweb.org/Sketcher_Workbench
https://wiki.freecadweb.org/Sketcher_ConstrainCoincident
https://wiki.freecadweb.org/Release_notes_1.0

**Comment by PaddleStroke** [comment]
@luzpaz can you please check and let me know if those changes are correct? Don't hesitate to modify if necessary.

**Comment by 0penBrain** [comment]
> @luzpaz can you please check and let me know if those changes are correct? Don't hesitate to modify if necessary.

What is important is that information about changes is pushed there. Then we have awesome people working on the wiki and they eventually can fix wording and so on (the good strategy IMO is that you learn from what these regulars fix).
Just 2 points:
 * Except in Release notes, think about tagging your changes with `{{Version|1.0}}` ;)
 * Be careful with translation tags (better not do it if you're unsure). For example in the page about Coincident constraint, you use a translation tag that is already used in the page, so that probably will be a mess. :)

**Pull request merged**
